### PR TITLE
Emulate msh2xdmf and make x-z meshing possible

### LINKED ISSUE #852: Emulate msh2xdmf functionality with options for 2-D meshing with specified plane and 3-D meshing.
## Description of issue / requirement to address

The existing open-source (LGPLv3) functionality available at https://github.com/floiseau/msh2xdmf is presently used but brings with it a number of downsides. It has to be installed locally (not via pip) and means we need to check if it was installed prior to calling it.

It also only handles 2-D meshing in the x-y plane, and a lot of our 2-D meshing is likely going to be required in the x-z plane. There is a workaround for this, but it is a little undesirable.

## Proposed solution

Given the minimal overhead, we agreed to emulate this functionality, providing the necessary modifications and sparing us from import checks.

Add a MSH to XMDF converter in `mesh/tools.py`, crediting the authors of the original repository (and linking to it).

## Alternative solutions

We could continue to use it as it or get in touch with the developers and propose modifications directly.

## Additional context
It's likely that any dependencies changes (e.g. moving to fenics-X) could bring in further modifications. LGPLv3 license could however be problematic.

@ivanmaione

### PR DESCRIPTION
**Pull request by CoronelBuendia**
## Linked Issues

Closes #852 

I'm going to wait for #846 to come in and rebase on that to catch all the plane changes and xdmf calls, etc.

## Description
Adds `msh_to_xmdf` and `import_mesh` to `mesh/tools.py`

x-z meshing now possibly directly through choice of `dimensions`. We discussed adding a `placement: BluemiraPlacement` argument for additional flexibility, but I couldn't immediately see how to extract BluemiraGeo objects from a .msh file so I've left it out of this PR so that we can at least get this (already rather large) PR in, which should replace the existing functionality with the additional benefit of direct x-z meshing.

## Interface Changes

`msh2xdmf` removed as a dependency
`msh2xdmf.msh2xdmf("Mesh", dim=2)` -> `msh_to_xdmf("Mesh", dimensions=(0,1))`
`msh2xdmf.import_mesh("Mesh", dim=2)` -> `import_mesh("Mesh", dimensions=(0,1))`

x-z meshing can be achieved with dimensions=(0, 2)
3-D meshing can be achieved with dimensions =(0,1,2)

## Checklist

I confirm that I have completed the following checks:

- [x] Tests run locally and pass `pytest tests --reactor`
- [x] Code quality checks run locally and pass `flake8` and `black .`
- [x] Documentation built locally and checked `sphinx-build -W documentation/source documentation/build`

### CODE REVIEW

**Comment by je-cook** [comment]
will this allow us to re upgrade gmsh?

**Comment by CoronelBuendia** [comment]
> will this allow us to re upgrade gmsh?

Uff, dunno.. @ivanmaione ?

**Comment by ivanmaione** [comment]
> > will this allow us to re upgrade gmsh?
> 
> Uff, dunno.. @ivanmaione ?

There should not be problem in upgrade gmsh, if they don't change the way in which physical groups and boundaries are identified. Same for meshio on the format of the cell data (we already got incompatibility with the last version of meshio).

**Comment by CoronelBuendia** [comment]
> > > will this allow us to re upgrade gmsh?
> > 
> > 
> > Uff, dunno.. @ivanmaione ?
> 
> There should not be problem in upgrade gmsh, if they don't change the way in which physical groups and boundaries are identified. Same for meshio on the format of the cell data (we already got incompatibility with the last version of meshio).

OK, but I think one for another PR once this is in.

File: examples/equilibria/fem_fixed_boundary/fem_magnetostatic_2D_single_coil.py [hunk 1]
```diff
@@ -132,16 +119,15 @@
...
 
-msh2xdmf.msh2xdmf("Mesh.msh", dim=2, directory=".")
+msh_to_xdmf("Mesh.msh", dimensions=(0, 2), directory=".")
 
-mesh, boundaries, subdomains, labels = msh2xdmf.import_mesh(
-    prefix="Mesh",
-    dim=2,
+mesh, boundaries, subdomains, labels = import_mesh(
+    "Mesh",
+    dimensions=(0, 2),
```
Comment (ivanmaione) [inline examples/equilibria/fem_fixed_boundary/fem_magnetostatic_2D_single_coil.py:130]: When the files xdmf have been already created, this code would produce the same result for `dimensions=(1,2)`, `dimensions=2`, or `dimensions=(0,1)`. This behavior can be misleading.  A possible solution could be to remove in the `import_mesh` function the possibility to generate the xdmf file:
```
if not os.path.exists(domain_file) or not os.path.exists(boundary_file):
        msh_to_xdmf(f"{file_prefix}.msh", dimensions=dimensions, directory=directory)
```
and to limit `dimensions` to the `mesh.topology().dim()` (i.e. remove `dimensions` from the function signature).

**Review by ivanmaione — CHANGES_REQUESTED** [review]
Just minor comments...

**Review by ivanmaione — APPROVED** [review]

**Comment by ivanmaione** [comment]
@CoronelBuendia, folder tests/mesh/__pycache__ should be removed.

**Comment by CoronelBuendia** [comment]
> @CoronelBuendia, folder tests/mesh/**pycache** should be removed.

Ah yeah, thanks. Think I must have force added the folder at some point forgetting about the pycache

**Review by ivanmaione — APPROVED** [review]

**Pull request merged**